contracts/rollover/CrossCurrencyRollover.sol

```diff
@@ -0,0 +1,662 @@
+        uint256 borrowerNoteId
+    ) internal view {
+        // ------------- Caller Validation -------------
+        address _borrower = IPromissoryNote(borrowerNote).ownerOf(borrowerNoteId);
+
+        if (_borrower != msg.sender) revert CCR_CallerNotBorrower();
+
+        // ------------- Rollover Terms Validation -------------
+        // currency must not be the same
+        if (sourceLoanTerms.payableCurrency == newLoanTerms.payableCurrency) {
+            revert CCR_SameCurrency(sourceLoanTerms.payableCurrency, newLoanTerms.payableCurrency);
+        }
+
+        // collateral address and id must be the same
+        if (sourceLoanTerms.collateralAddress != newLoanTerms.collateralAddress || sourceLoanTerms.collateralId != newLoanTerms.collateralId) {
+            revert CCR_CollateralMismatch(
+                sourceLoanTerms.collateralAddress,
+                sourceLoanTerms.collateralId,
+                newLoanTerms.collateralAddress,
+                newLoanTerms.collateralId
+            );
+        }
+
+        // ------------- New LoanTerms Validation -------------
+        originationHelpers.validateLoanTerms(newLoanTerms);
+    }
+
+    // ========================================= HELPERS ================================================
+    /**
+     * @notice Helper function to distribute funds based on the rollover amounts. A flash loan must be
+     *         initiated to repay the old loan.
+     *
+     * @param oldLoanId                 The ID of the original loan to be rolled over.
+     * @param oldLoanData               The loan data of the original loan.
+     * @param newPrincipalAmount        The principal amount of the new loan.
+     * @param newCurrency               The currency of the new loan.
+     * @param borrower_                 The address of the borrower.
+     * @param lender                    The address of the new lender.
+     *
+     * @return amounts                  The rollover amounts.
+     * @return feeSnapshot              A snapshot of current lending fees.
+     * @return repayAmount              The amount needed to repay the old loan.
+     */
+    function _rollover(
+        uint256 oldLoanId,
+        LoanLibrary.LoanData memory oldLoanData,
+        uint256 newPrincipalAmount,
+        address newCurrency,
+        address borrower_,
+        address lender,
+        uint24 poolFee
+    ) internal nonReentrant returns (
+        OriginationLibrary.RolloverAmounts memory amounts,
+        LoanLibrary.FeeSnapshot memory feeSnapshot,
+        uint256 repayAmount
+    ) {
+        address oldLender = ILoanCore(loanCore).lenderNote().ownerOf(oldLoanId);
+        IERC20 payableCurrency = IERC20(oldLoanData.terms.payableCurrency);
+
+        // get fee snapshot from fee controller
+        (feeSnapshot) = feeController.getFeeSnapshot();
+
+        // Calculate settle amounts
+        (amounts, repayAmount) = _calculateCurrencyRolloverAmounts(
```

Comment:
Thinking here...the flow makes sense but still lacks slippage protection. In case the price moves "against" the borrower before the time has submitted, the borrower might be surprised at the amount of tokens they need to repay the rollover. In fact, since according to line 580 you're always calling `_swapExactInputSingle` with a `0` for `amountOutMinimum`, the borrower will _always_ get slipped, which is bad.

So to prevent this I think you will still need a `minPrice` parameter or a `minAmountOut` parameter, to be provided by the borrower, that will prevent the rollover from going through in cases where they would end up paying a higher price than anticipated. In terms of input parameters, you could probably combine this with `poolFee` into a struct called something like `SwapParameters`. And to be double-safe you would also want to convert the `minPrice` into an `amountOutMinimum` that gets passed into `opData` then into `_swapExactInputSingle`. 

Then, if you had a `minPrice`, you could rework things a little bit, such that you fetch the current price earlier on in the flow (before the signature recovery), then revert if `price < minPrice`. Then, `_rollover` and `_calculateCurrencyRolloverAmounts` could take the price as arguments directly, as opposed to `poolFee`.

Alternatively, you could use `minAmountOut` as a parameter for slippage protection, and validate it against `amountOut` at the end of `_swapExactInputSingle`. I like this a bit better because it's more straightforward; when we have that `amountOut` variable, we have a direct guarantee that the contract received exactly that many tokens. With `minPrice`, the amount is more indirect, and is only really checked pre-swap, not post-swap.

Lastly, once you implement slippage protection, that will also affect the settlement value vis-a-vis the borrower. You handle the case where we have _too many_ tokens well, with the `remainingFunds` variable, but since we don't know what `amountOut` will be in `_calculateCurrencyRolloverAmounts`, we don't know exactly what `amounts.needFromBorrower` for be.

If you move that same calculation down to the post-swap stage, you can collect it at the stage as well; since you only need the borrower's funds to repay the flash loan, which hasn't been done yet. You could have something like that looks like:

```

        // any amount left over after repaying the flash loan is sent to the borrower
        // it does not == opData.rolloverAmounts.amountToBorrower because of slippage
        uint256 remainingFunds = swappedAmount - amount;

        if (swappedAmount > amount) {
            uint256 remainingFunds = swappedAmount - amount;

            // if funds remain, send the difference to the borrower
            asset.safeTransfer(borrower, remainingFunds);
        } else if (swappedAmount < amount) {
           uint256 needFromBorrower = amount - swappedAmount;

            // Borrower owes from old loan
            payableCurrency.safeTransferFrom(borrower_, address(this), amounts.needFromBorrower);
        }
```

You were almost there already, since you mentioned that `remainingFunds != opData.rolloverAmounts.amountToBorrower` - it's just that the same thing would be true for `needFromBorrower` as well, so we should handle it later as well.